In [3]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

In [7]:
df_cafe_dirty = pd.read_csv("/Users/maitanelopezsanchez/mi_proyecto-1/Datasets/dirty_cafe_sales.csv")

In [10]:
df_cafe_dirty

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
2       TXN_4271903    Cookie        4            1.0       ERROR   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9995    TXN_7672686    Coffee        2            2.0         4.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9997    TXN_5255387    Coffee        4            2.0         8.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07  

[10000 rows x 8 columns]

# Primera Parte
## Carga, Exploración, Limpieza y Visualización Básica de un Dataset

In [60]:
df_cafe_dirty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          10000 non-null  object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [61]:
df_cafe_dirty.shape

(10000, 8)

In [62]:
df_cafe_dirty.items()

<generator object DataFrame.items at 0x10e6398a0>

In [63]:
df_cafe_dirty.head()

Transaction ID    Item Quantity Price Per Unit Total Spent  Payment Method  \
0    TXN_1961373  Coffee        2            2.0         4.0     Credit Card   
1    TXN_4977031    Cake        4            3.0        12.0            Cash   
2    TXN_4271903  Cookie        4            1.0       ERROR     Credit Card   
3    TXN_7034554   Salad        2            5.0        10.0         UNKNOWN   
4    TXN_3160411  Coffee        2            2.0         4.0  Digital Wallet   

   Location Transaction Date  
0  Takeaway       2023-09-08  
1  In-store       2023-05-16  
2  In-store       2023-07-19  
3   UNKNOWN       2023-04-27  
4  In-store       2023-06-11

In [64]:
df_cafe_dirty.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='object')

In [ ]:
df_cafe_dirty.Item.unique()
#Hay 3 tipos diferentes de valores nulos, hay que unificarlos, transformarlos o eliminarlos de la forma más coherente.

array(['Coffee', 'Cake', 'Cookie', 'Salad', 'Smoothie', 'UNKNOWN',
       'Sandwich', nan, 'ERROR', 'Juice', 'Tea'], dtype=object)

In [66]:
df_cafe_dirty["Transaction ID"].value_counts()
# Al ver que la longitud es de 10.000 podemos observar que coincide con el lenght y no hay "Transaction ID" repetidos.

Transaction ID
TXN_1961373    1
TXN_4831525    1
TXN_1228927    1
TXN_6486912    1
TXN_3447069    1
              ..
TXN_5680238    1
TXN_7766134    1
TXN_2617257    1
TXN_8993132    1
TXN_6170729    1
Name: count, Length: 10000, dtype: int64

In [67]:
df_cafe_dirty['Quantity'].isna().any()
# No hay valores nulos en la columna Quantity ya que nos devuelve un False

np.False_

In [ ]:
df_cafe_dirty["Item"].isna().any()
#Efectivamente hay valores nulos en la columna Item.

np.True_

In [93]:
df_cafe_dirty["Price Per Unit"].isna().any()

np.True_

In [99]:
df_cafe_dirty["Transaction Date"].isna().any()

np.True_

In [107]:
df_cafe_dirty["Quantity"].isna().any()
#Quantity y Transaction ID son las dos unicas columnas sin valores nulos.

np.False_

In [108]:
df_cafe_dirty["Payment Method"].unique() 
#3 formas de pago y 3 tipos de valores nulos

array(['Credit Card', 'Cash', 'UNKNOWN', 'Digital Wallet', 'ERROR', nan],
      dtype=object)

## Comenzando la limpieza

In [109]:
df_cafe_dirty.groupby("Item")["Price Per Unit"].first()


Item
Cake        3.0
Coffee      2.0
Cookie      1.0
ERROR       1.5
Juice       3.0
Salad       5.0
Sandwich    4.0
Smoothie    4.0
Tea         1.5
UNKNOWN     3.0
Name: Price Per Unit, dtype: object

Tras observar el valor de cada Item, podemos ver como hay una coincidencia con el Price Per Unit del Item "ERROR"

In [110]:
df_cafe_dirty["Price Per Unit"].value_counts()

Price Per Unit
3.0        2429
4.0        2331
2.0        1227
5.0        1204
1.0        1143
1.5        1133
ERROR       190
UNKNOWN     164
Name: count, dtype: int64

In [111]:
df_cafe_dirty[df_cafe_dirty["Price Per Unit"].isin(["Unknown", "ERROR"])]


Transaction ID      Item Quantity Price Per Unit Total Spent  \
68      TXN_8427104     Salad        2          ERROR        10.0   
196     TXN_3800182       Tea        2          ERROR         3.0   
212     TXN_9089898     Salad        3          ERROR        15.0   
213     TXN_8587583      Cake        1          ERROR         3.0   
289     TXN_3495950       NaN        4          ERROR         6.0   
...             ...       ...      ...            ...         ...   
9533    TXN_4434185  Smoothie        5          ERROR        20.0   
9605    TXN_9397870      Cake        2          ERROR         6.0   
9649    TXN_2509803    Coffee        1          ERROR         2.0   
9717    TXN_3334632       NaN        1          ERROR         2.0   
9753    TXN_3507013     Juice        1          ERROR         3.0   

      Payment Method  Location Transaction Date  
68               NaN  In-store       2023-10-27  
196             Cash  In-store       2023-10-26  
212      Credit Card       NaN       2023-01-11  
213              NaN       NaN       2023-10-01  
289      Credit Card  In-store       2023-02-19  
...              ...       ...              ...  
9533     Credit Card  Takeaway       2023-04-30  
9605             NaN       NaN       2023-05-23  
9649         UNKNOWN       NaN       2023-02-22  
9717     Credit Card  Takeaway       2023-11-20  
9753  Digital Wallet  In-store       2023-01-16  

[190 rows x 8 columns]

In [112]:
df_cafe_dirty[df_cafe_dirty["Price Per Unit"].astype(str).str.contains("[A-Za-z]", regex=True)]
#De esta forma podemos ver si alguno de los valores del Price Per Unit tenía algún valor como "$" u otros.


Transaction ID      Item Quantity Price Per Unit Total Spent  \
56      TXN_3578141      Cake        5            NaN        15.0   
65      TXN_4987129  Sandwich        3            NaN         NaN   
68      TXN_8427104     Salad        2          ERROR        10.0   
85      TXN_8035512       Tea        3            NaN         4.5   
104     TXN_7447872     Juice        2            NaN         6.0   
...             ...       ...      ...            ...         ...   
9924    TXN_5981429     Juice        2            NaN         6.0   
9926    TXN_2464706      Cake        4        UNKNOWN        12.0   
9961    TXN_2153100       Tea        2        UNKNOWN         3.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   

      Payment Method  Location Transaction Date  
56               NaN  Takeaway       2023-06-27  
65               NaN  In-store       2023-10-20  
68               NaN  In-store       2023-10-27  
85              Cash   UNKNOWN       2023-10-29  
104              NaN       NaN              NaN  
...              ...       ...              ...  
9924  Digital Wallet       NaN       2023-12-24  
9926  Digital Wallet  Takeaway       2023-11-09  
9961            Cash       NaN       2023-12-29  
9996  Digital Wallet       NaN       2023-06-02  
9998  Digital Wallet       NaN       2023-12-02  

[533 rows x 8 columns]

In [113]:
#A continuación, vamos a eliminar las filas que contentan valores nulos como son "NaN", "ERROR" o "UNKNOWN"
# tanto en la la columna Item como en la columna Price Per Unit, ya que, no nos aportan ninguna información.

In [114]:
invalid_values = ["ERROR", "UNKNOWN"]


In [118]:
#Función que elimine las filas donde AMBOS valores son invalid_values
mask_invalid_both = (
    (df_cafe_dirty["Item"].isin(invalid_values) | df_cafe_dirty["Item"].isna()) &
    (df_cafe_dirty["Price Per Unit"].isin(invalid_values) | df_cafe_dirty["Price Per Unit"].isna())
)


In [120]:
df_cafe_dirty = df_cafe_dirty[~mask_invalid_both]


In [121]:
df_cafe_dirty[
    (df_cafe_dirty["Item"].isin(invalid_values) | df_cafe_dirty["Item"].isna()) &
    (df_cafe_dirty["Price Per Unit"].isin(invalid_values) | df_cafe_dirty["Price Per Unit"].isna())
]



Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []

In [124]:
df_cafe_dirty.shape

(9946, 8)

Hemos eliminado y guardado la primera fase de la limpieza, podemos ver que el tamaño de la tabla de ha reducido de 10.000 filas a 9.946.

El siguiente paso que quiero dar es transormar en valores tangibles aquellos invalid_values donde sabemos el precio por unidad y no el Item o a la inversa. Por ejemplo, sabemos que un Coffee vale 2.0 y la Cookie 1.0. Y al revés, donde el precio sea 1.0 y el Item un invalid_value poder decir que es una Cookie. 
    Nota: En algunos casos se complica ya que el precio de la Cake y del Juice es el mismo.

In [ ]:
df_cafe_dirty.groupby("Item")["Price Per Unit"].first()

Item
Cake        3.0
Coffee      2.0
Cookie      1.0
ERROR       1.5
Juice       3.0
Salad       5.0
Sandwich    4.0
Smoothie    4.0
Tea         1.5
UNKNOWN     3.0
Name: Price Per Unit, dtype: object

Como podemos observar, los Item: "ERROR" tienen un Price Per Unit de 1.5. El único Item que tiene ese Precio es el Té, por ello podemos transformar los Item que sean ERROR a Té.

In [131]:
df_cafe_dirty.loc[
    (df_cafe_dirty["Item"] == "ERROR") & (df_cafe_dirty["Price Per Unit"] == "1.5"),
    "Item"
] = "Tea"


In [133]:
df_cafe_dirty[df_cafe_dirty["Price Per Unit"] == "1.5"]


Transaction ID Item Quantity Price Per Unit Total Spent  Payment Method  \
14      TXN_8915701  Tea        2            1.5         3.0             NaN   
42      TXN_6650263  Tea        2            1.5     UNKNOWN             NaN   
70      TXN_2621580  Tea        2            1.5         3.0            Cash   
78      TXN_8735480  Tea        5            1.5         7.5            Cash   
90      TXN_8976658  Tea        2            1.5         3.0     Credit Card   
...             ...  ...      ...            ...         ...             ...   
9955    TXN_9187008  Tea        4            1.5         6.0           ERROR   
9965    TXN_5846013  Tea        2            1.5         3.0  Digital Wallet   
9974    TXN_8076061  Tea        4            1.5         6.0            Cash   
9978    TXN_4302199  Tea        3            1.5         4.5             NaN   
9980    TXN_6796890  Tea        4            1.5         6.0         UNKNOWN   

      Location Transaction Date  
14    In-store       2023-03-21  
42    Takeaway       2023-01-10  
70    In-store       2023-03-22  
78         NaN       2023-06-02  
90    In-store       2023-08-16  
...        ...              ...  
9955       NaN       2023-09-16  
9965       NaN       2023-12-20  
9974  In-store       2023-04-01  
9978       NaN       2023-02-16  
9980       NaN       2023-08-24  

[1133 rows x 8 columns]

Segundo paso, transformar aquellos Item con invalid_values en un valor cuando se cumpla lo siguiente: 
Coffee → 2.0
Cookie → 1.0
Salad → 5.0
Tea → 1.5
En el caso de Items con Price Per Unit iguales no se podrá realizar este cambio.

In [134]:
mapping = {
    "2.0": "Coffee",
    "1.0": "Cookie",
    "5.0": "Salad",
    "1.5": "Tea"
}

In [135]:
for price_str, item_name in mapping.items():
    df_cafe_dirty.loc[
        (df_cafe_dirty["Item"].isin(invalid_values)) &
        (df_cafe_dirty["Price Per Unit"] == price_str),
        "Item"
    ] = item_name

In [137]:
#Verificación
df_cafe_dirty[df_cafe_dirty["Item"].isin(mapping.values())]


Transaction ID    Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373  Coffee        2            2.0         4.0   
2       TXN_4271903  Cookie        4            1.0       ERROR   
3       TXN_7034554   Salad        2            5.0        10.0   
4       TXN_3160411  Coffee        2            2.0         4.0   
10      TXN_2548360   Salad        5            5.0        25.0   
...             ...     ...      ...            ...         ...   
9982    TXN_8567525  Cookie        2            1.0         2.0   
9990    TXN_1538510  Coffee        5            2.0        10.0   
9995    TXN_7672686  Coffee        2            2.0         4.0   
9997    TXN_5255387  Coffee        4            2.0         8.0   
9998    TXN_7695629  Cookie        3            NaN         3.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
10              Cash  Takeaway       2023-11-07  
...              ...       ...              ...  
9982             NaN  Takeaway       2023-12-30  
9990  Digital Wallet       NaN       2023-05-22  
9995             NaN   UNKNOWN       2023-08-30  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  

[4814 rows x 8 columns]

Tercer paso: Transformar los Price Per Unit que contengan invalid_values.

In [138]:
#Generamos un mapping con los precios de cada Item. Cake - 3.0, etc.
price_mapping = {
    "Cake": "3.0",
    "Coffee": "2.0",
    "Cookie": "1.0",
    "Juice": "3.0",
    "Salad": "5.0",
    "Sandwich": "4.0",
    "Smoothie": "4.0",
    "Tea": "1.5"
}

In [139]:
for item_name, price_str in price_mapping.items():
    df_cafe_dirty.loc[
        (df_cafe_dirty["Price Per Unit"].isin(invalid_values)) &
        (df_cafe_dirty["Item"] == item_name),
        "Price Per Unit"
    ] = price_str


In [ ]:
df_cafe_dirty[df_cafe_dirty["Price Per Unit"].isin(invalid_values)]
#Devuelve una tabla vacía por lo que ya no hay valores nulos en la columna Price Per Unit.


Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []

In [141]:
df_cafe_dirty[df_cafe_dirty["Item"] == "UNKNOWN"]


Transaction ID     Item Quantity Price Per Unit Total Spent  \
6       TXN_4433211  UNKNOWN        3            3.0         9.0   
36      TXN_6855453  UNKNOWN        4            3.0        12.0   
91      TXN_5455936  UNKNOWN        5            3.0        15.0   
153     TXN_6541415  UNKNOWN  UNKNOWN            3.0        12.0   
165     TXN_3226832  UNKNOWN        5            4.0        20.0   
...             ...      ...      ...            ...         ...   
9727    TXN_3562418  UNKNOWN        2            4.0         8.0   
9763    TXN_7652830  UNKNOWN        2            3.0         6.0   
9836    TXN_9162296  UNKNOWN        3            4.0        12.0   
9946    TXN_8807600  UNKNOWN        1            4.0         4.0   
9994    TXN_7851634  UNKNOWN        4            4.0        16.0   

      Payment Method  Location Transaction Date  
6              ERROR  Takeaway       2023-10-06  
36               NaN  In-store       2023-07-17  
91               NaN  In-store       2023-10-28  
153             Cash  In-store       2023-11-25  
165             Cash   UNKNOWN       2023-09-04  
...              ...       ...              ...  
9727  Digital Wallet   UNKNOWN       2023-08-13  
9763             NaN       NaN       2023-08-15  
9836            Cash  In-store       2023-05-10  
9946            Cash  Takeaway       2023-09-24  
9994             NaN       NaN       2023-01-08  

[147 rows x 8 columns]

Debido a que los UNKNOWN que quedan tienen precio de 3.0 y de 4.0 y ese precio puede ser de 2 Items diferentes, considero que la mejor opción es eliminarlas ya que solo son 147 de 10.000 filas.

In [145]:
df_cafe_dirty = df_cafe_dirty[df_cafe_dirty["Item"] != "UNKNOWN"]

In [147]:
df_cafe_dirty[df_cafe_dirty["Item"] == "UNKNOWN"]
#Verificado que se han eliminado correctamente.

Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []

In [156]:
df_cafe_dirty

Empty DataFrame
Columns: [Transaction ID, Item, Quantity, Price Per Unit, Total Spent, Payment Method, Location, Transaction Date]
Index: []